### ISSUE TITLE
torch.equal not give the same result on GPU vs on CPU

### ISSUE THREAD

**Issue by cheyennee** [pytorch/pytorch#83314]
### 🐛 Describe the bug

```
results = dict()
import torch
arg_1 = torch.randint(-32768,8,[2], dtype=torch.int64)
arg_2 = torch.randint(-1,32,[4], dtype=torch.int8)
try:
  results["res_cpu"] = torch.equal(arg_1,arg_2,)
except Exception as e:
  results["err_cpu"] = "ERROR:"+str(e)
arg_1 = arg_1.clone().cuda()
arg_2 = arg_2.clone().cuda()
try:
  results["res_gpu"] = torch.equal(arg_1,arg_2,)
except Exception as e:
  results["err_gpu"] = "ERROR:"+str(e)
print(results)
```
The result is as follows:
{'err_cpu': "ERROR:Expected object of scalar type __int64 but got scalar type signed char for argument 'other'", 'res_gpu': False}.

However, it's expected that cpu and gpu output the same result, not one outputs error message, and another runs successfully

### Versions

pytorch: 1.8.1
python: 3.8.3
os: win11

cc @VitalyFedyunin

**Comment by malfet** [comment]
Still reproducible on nightly and documentation is a bit vague, what is the expected output. I.e. 
`torch.equal(torch.tensor([1, 2]), torch.tensor([1, 2], dtype=torch.float))` raises exception, but I guess it should not

**Comment by ezyang** [comment]
BTW, torch.equal should be deprecated. You should be using allclose instead

**Comment by vadimkantorov** [comment]
Is torch.equal alias for operator `==`?

**Comment by ezyang** [comment]
no, it returns a bool

**Comment by ILoveSE** [comment]
Seems to be affected by the same underlying implementation, the same issue occurs with `torch.Tensor.equal` on version 1.12.0 and returns `{'err_cpu': "ERROR: Expected object of scalar type long int but got scalar type signed char for argument 'other'", 'res_gpu': False}`.
Fortunately, **this misbehavior has been fixed in 1.13.0, this comment is only for reporting this behavior, not asking for a backported fix**.

```python
results = dict()
import torch
arg_1 = torch.randint(-32768,8,[2], dtype=torch.int64)
arg_2 = torch.randint(-1,32,[4], dtype=torch.int8)
try:
  results["res_cpu"] = torch.Tensor.equal(arg_1,arg_2,)
except Exception as e:
  results["err_cpu"] = "ERROR:"+str(e)
arg_1 = arg_1.clone().cuda()
arg_2 = arg_2.clone().cuda()
try:
  results["res_gpu"] = torch.Tensor.equal(arg_1,arg_2,)
except Exception as e:
  results["err_gpu"] = "ERROR:"+str(e)
print(results)
```
